File: Sources/vibetrail/Commands.swift
import ArgumentParser
import Foundation
import VibeTrailCore

struct ProjectsCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "projects", abstract: "List projects across all providers (F1).")

    @Flag(name: .long, help: "Emit JSON.") var json = false

    func run() throws {
        try CLIEnvironment.run {
            let projects = try CLIEnvironment.store().projects()
            if json {
                print(try VibeTrailJSON.encode(projects))
                return
            }
            if projects.isEmpty {
                print("No sessions found.")
                return
            }
            for project in projects {
                let marker = project.exists ? " " : "!"
                let path = Format.abbreviatePath(project.realPath)
                let providers = project.providers.sorted().joined(separator: ",")
                var line = "\(marker) \(Format.pad(path, 52)) \(Format.pad("\(project.sessionCount)", 5)) "
                line += "\(Format.pad(Format.relativeTime(project.lastActive), 12)) \(providers)"
                if let prompt = project.lastPrompt {
                    line += "  \(Format.truncate(prompt, to: 60))"
                }
                print(line)
            }
        }
    }
}

struct SessionsCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "sessions", abstract: "List sessions of a project, newest first (F2).")

    @Argument(help: "Project path, or a unique directory-name suffix.") var project: String
    @Option(name: .customShort("n"), help: "Maximum sessions to list.") var limit: Int = 20
    @Option(name: .long, help: "Restrict to one provider id.") var provider: String?
    @Flag(name: .long, help: "Emit JSON.") var json = false

    func run() throws {
        try CLIEnvironment.run {
            let store = CLIEnvironment.store()
            let projectPath = try store.resolveProject(project)
            let sessions = try store.sessions(inProject: projectPath, providerId: provider, limit: limit)
            if json {
                print(try VibeTrailJSON.encode(sessions))
                return
            }
            if sessions.isEmpty {
                print("No sessions in \(Format.abbreviatePath(projectPath)).")
                return
            }
            for session in sessions {
                var line = "\(session.nativeId.prefix(8))  \(Format.pad(Format.relativeTime(session.mtime), 12)) "
                line += "\(Format.pad("\(session.messageCount) msg", 9)) \(Format.pad(Format.duration(session.duration), 6)) "
                if let branch = session.gitBranch {
                    line += "\(Format.pad(branch, 14)) "
                }
                line += Format.truncate(session.title, to: 70)
                print(line)
            }
        }
    }
}

struct SearchCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "search", abstract: "Full-text search across providers and projects (F4).")

    @Argument(help: "Literal text to search for.") var query: String
    @Option(name: [.customShort("p"), .long], help: "Restrict to one project.") var project: String?
    @Option(name: .long, help: "Restrict to one provider id.") var provider: String?
    @Flag(name: .long, help: "Emit JSON.") var json = false

    func run() throws {
        try CLIEnvironment.run {
            let store = CLIEnvironment.store()
            let projectPath = try project.map { try store.resolveProject($0) }
            let engine = try CLIEnvironment.searchEngine()
            let hits = try engine.search(query, scope: SearchScope(projectPath: projectPath,
                                                                   providerId: provider))
            if json {
                print(try VibeTrailJSON.encode(hits))
                return
            }
            if hits.isEmpty {
                print("No matches for \"\(query)\".")
                return
            }
            // F4: results aggregated per session.
            var order: [String] = []
            var grouped: [String: [SearchHit]] = [:]
            for hit in hits {
                if grouped[hit.sessionId] == nil { order.append(hit.sessionId) }
                grouped[hit.sessionId, default: []].append(hit)
            }
            for sessionId in order {
                let sessionHits = grouped[sessionId]!
                let first = sessionHits[0]
                print("\(first.nativeSessionId.prefix(8))  \(Format.abbreviatePath(first.projectPath))  [\(first.providerId)]")
                for hit in sessionHits.prefix(5) {
                    print("    \(Format.truncate(hit.snippet, to: 140))")
                }
                if sessionHits.count > 5 {
                    print("    … \(sessionHits.count - 5) more matches")
                }
            }
        }
    }
}

struct ShowCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "show", abstract: "Show one session as outline (default) or full transcript (F3).")

    @Argument(help: "Session id (or unique prefix).") var sessionId: String
    @Flag(name: .long, help: "Outline view (default).") var outline = false
    @Flag(name: .long, help: "Full transcript.") var full = false
    @Flag(name: .long, help: "Emit JSON.") var json = false

    func run() throws {
        try CLIEnvironment.run {
            if outline && full {
                throw VibeTrailError.usage("--outline and --full are mutually exclusive")
            }
            let store = CLIEnvironment.store()
            let (provider, raw) = try store.resolveSession(sessionId)
            if full {
                let session = try provider.parse(raw)
                if json {
                    print(try VibeTrailJSON.encode(session))
                    return
                }
                printHeader(session.summary)
                for message in session.messages {
                    printFull(message)
                }
            } else {
                let stubs = try provider.outline(raw)
                if json {
                    print(try VibeTrailJSON.encode(stubs))
                    return
                }
                printHeader(try provider.summarize(raw))
                for stub in stubs {
                    let index = Format.pad("\(stub.index)", 4)
                    print("\(index) \(Format.roleIcon(stub.role)) \(Format.truncate(stub.preview, to: 110))")
                }
            }
        }
    }

    private func printHeader(_ summary: SessionSummary) {
        print("\(summary.title)")
        var meta = "\(summary.id) · \(Format.abbreviatePath(summary.projectPath)) · "
        meta += "\(summary.messageCount) messages · \(Format.relativeTime(summary.mtime))"
        if let branch = summary.gitBranch { meta += " · \(branch)" }
        print(meta)
        print(String(repeating: "─", count: 80))
    }

    private func printFull(_ message: Message) {
        print("\(Format.roleIcon(message.role)) [\(message.role.rawValue)]")
        for block in message.blocks {
            switch block {
            case .text(let text):
                print(text)
            case .toolUse(let name, _):
                print("  ⚙ tool_use: \(name)")
            case .toolResult(let summary, let truncated):
                let firstLine = summary.split(whereSeparator: \.isNewline).first ?? ""
                print("  → tool_result: \(Format.truncate(String(firstLine), to: 100))\(truncated ? " …" : "")")
            case .thinking:
                print("  ✳ thinking (collapsed)")
            }
        }
        print("")
    }
}

struct ResumeCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "resume", abstract: "Resume a session in its project directory (F5).")

    @Argument(help: "Session id (or unique prefix).") var sessionId: String

    func run() throws {
        try CLIEnvironment.run {
            let spec = try CLIEnvironment.store().resumeSpec(for: sessionId)
            try CLIResumer().resume(spec)
        }
    }
}

struct OpenCommand: ParsableCommand {
    static let configuration = CommandConfiguration(
        commandName: "open", abstract: "Open the VibeTrail GUI app.")

    @Argument(help: "Optional project to open at.") var project: String?

    func run() throws {
        try CLIEnvironment.run {
            let process = Process()
            process.executableURL = URL(fileURLWithPath: "/usr/bin/open")
            process.arguments = ["-a", "VibeTrail"] + (project.map { [$0] } ?? [])
            try? process.run()
            process.waitUntilExit()
            if process.terminationStatus != 0 {
                throw VibeTrailError.data("VibeTrail.app not found; build it from the VibeTrailApp target")
            }
        }
    }
}

File: crates/vibetrail-cli/src/commands.rs
use std::collections::HashMap;
use std::os::unix::process::CommandExt;

use vibetrail_core::{
    search_store, ContentBlock, Error, Message, Result, Scope, SearchHit, Session, SessionStore,
    SessionSummary,
};

use crate::format;

fn to_json<T: serde::Serialize>(value: &T) -> Result<String> {
    serde_json::to_string_pretty(value).map_err(|e| Error::Data(format!("JSON encoding: {e}")))
}

pub fn projects(store: &SessionStore, json: bool) -> Result<()> {
    let projects = store.projects()?;
    if json {
        println!("{}", to_json(&projects)?);
        return Ok(());
    }
    if projects.is_empty() {
        println!("No sessions found.");
        return Ok(());
    }
    for project in projects {
        let marker = if project.exists { " " } else { "!" };
        let providers: Vec<&str> = project.providers.iter().map(String::as_str).collect();
        let mut line = format!(
            "{marker} {} {} {} {}",
            format::pad(&format::abbreviate_path(&project.real_path), 52),
            format::pad(&project.session_count.to_string(), 5),
            format::pad(&format::relative_time(&project.last_active), 12),
            providers.join(","),
        );
        if let Some(prompt) = &project.last_prompt {
            line.push_str(&format!("  {}", format::truncate(prompt, 60)));
        }
        println!("{line}");
    }
    Ok(())
}

pub fn sessions(
    store: &SessionStore,
    project: &str,
    limit: usize,
    provider: Option<&str>,
    json: bool,
) -> Result<()> {
    let project_path = store.resolve_project(project)?;
    let mut handles = store.session_handles(&project_path, provider)?;
    handles.truncate(limit);
    let sessions = store.summarize_handles(&handles);
    if json {
        println!("{}", to_json(&sessions)?);
        return Ok(());
    }
    if sessions.is_empty() {
        println!("No sessions in {}.", format::abbreviate_path(&project_path));
        return Ok(());
    }
    // Resume/fork chain markers: ↳ when the parent is in this project.
    let known: std::collections::HashSet<&str> =
        handles.iter().map(|h| h.native_id.as_str()).collect();
    let parent_of: std::collections::HashMap<&str, &str> = handles
        .iter()
        .filter_map(|h| {
            let parent = h.parent_native_id.as_deref()?;
            known
                .contains(parent)
                .then_some((h.native_id.as_str(), parent))
        })
        .collect();
    for session in sessions {
        let short_id: String = session.native_id.chars().take(8).collect();
        let chain = match parent_of.get(session.native_id.as_str()) {
            Some(parent) => format!("↳ {} ", &parent[..8.min(parent.len())]),
            None => String::new(),
        };
        let mut line = format!(
            "{short_id}  {} {} {} ",
            format::pad(&format::relative_time(&session.mtime), 12),
            format::pad(&format!("{} msg", session.message_count), 9),
            format::pad(&format::duration(session.duration), 6),
        );
        if let Some(branch) = &session.git_branch {
            line.push_str(&format!("{} ", format::pad(branch, 14)));
        }
        line.push_str(&chain);
        line.push_str(&format::truncate(&session.title, 70));
        println!("{line}");
    }
    Ok(())
}

pub fn search(
    store: &SessionStore,
    query: &str,
    project: Option<&str>,
    provider: Option<&str>,
    json: bool,
) -> Result<()> {
    let project_path = project.map(|p| store.resolve_project(p)).transpose()?;
    let scope = Scope {
        project_path,
        provider_id: provider.map(String::from),
    };
    let hits = search_store(store, query, &scope)?;
    if json {
        println!("{}", to_json(&hits)?);
        return Ok(());
    }
    if hits.is_empty() {
        println!("No matches for \"{query}\".");
        return Ok(());
    }
    // F4: results aggregated per session.
    let mut order: Vec<String> = Vec::new();
    let mut grouped: HashMap<String, Vec<&SearchHit>> = HashMap::new();
    for hit in &hits {
        grouped
            .entry(hit.session_id.clone())
            .or_insert_with(|| {
                order.push(hit.session_id.clone());
                Vec::new()
            })
            .push(hit);
    }
    for session_id in order {
        let session_hits = &grouped[&session_id];
        let first = session_hits[0];
        let short_id: String = first.native_session_id.chars().take(8).collect();
        println!(
            "{short_id}  {}  [{}]",
            format::abbreviate_path(&first.project_path),
            first.provider_id
        );
        for hit in session_hits.iter().take(5) {
            println!("    {}", format::truncate(&hit.snippet, 140));
        }
        if session_hits.len() > 5 {
            println!("    … {} more matches", session_hits.len() - 5);
        }
    }
    Ok(())
}

/// Custom-agent roster for a project: project-level definitions plus the
/// user-global ones, one line each (frontmatter description only — bodies
/// are system prompts, use --json for full content).
pub fn agents(store: &SessionStore, project: &str, json: bool) -> Result<()> {
    let project_path = store.resolve_project(project)?;
    let defs = store.project_agents(&project_path);
    if json {
        println!("{}", to_json(&defs)?);
        return Ok(());
    }
    if defs.is_empty() {
        println!(
            "No custom agents for {}.",
            format::abbreviate_path(&project_path)
        );
        return Ok(());
    }
    for def in defs {
        let mut line = format!(
            "{} {} [{}]",
            format::pad(&def.name, 24),
            format::pad(&def.scope, 8),
            def.provider_id
        );
        if let Some(model) = &def.model {
            line.push_str(&format!(" {model}"));
        }
        println!("{line}");
        if let Some(description) = &def.description {
            println!("    {}", format::truncate(description, 120));
        }
    }
    Ok(())
}

/// Handoff (TECH_SPEC §14): print the template prompt for continuing this
/// session in another agent. Plain print is pipe-friendly by design
/// (`vibetrail handoff <id> | pbcopy`).
pub fn handoff(store: &SessionStore, session_id: &str, json: bool) -> Result<()> {
    let (provider, raw) = store.resolve_session(session_id)?;
    let session = provider.parse(&raw)?;
    let capsule = vibetrail_core::HandoffCapsule::from_session(&session);
    if json {
        println!("{}", to_json(&capsule)?);
    } else {
        print!("{}", capsule.prompt());
    }
    Ok(())
}

/// F7: agent-persisted project memory, read-only, grouped as the providers
/// return it (each provider orders its own docs, index first).
pub fn memory(store: &SessionStore, project: &str, json: bool) -> Result<()> {
    let project_path = store.resolve_project(project)?;
    let docs = store.project_memory(&project_path);
    if json {
        println!("{}", to_json(&docs)?);
        return Ok(());
    }
    if docs.is_empty() {
        println!(
            "No agent memory for {}.",
            format::abbreviate_path(&project_path)
        );
        return Ok(());
    }
    for doc in docs {
        let mut header = format!("── [{}] {}", doc.provider_id, doc.name);
        if let Some(doc_type) = &doc.doc_type {
            header.push_str(&format!(" ({doc_type})"));
        }
        println!("{header}");
        if let Some(description) = &doc.description {
            println!("   {description}");
        }
        println!("{}", doc.content.trim_end());
        println!();
    }
    Ok(())
}

pub fn show(store: &SessionStore, session_id: &str, full: bool, json: bool) -> Result<()> {
    let (provider, raw) = store.resolve_session(session_id)?;
    if full {
        let session = provider.parse(&raw)?;
        if json {
            println!("{}", to_json(&session)?);
            return Ok(());
        }
        print_header(&session.summary);
        print_extensions(&session);
        for message in &session.messages {
            print_full(message);
        }
    } else {
        let stubs = provider.outline(&raw)?;
        if json {
            println!("{}", to_json(&stubs)?);
            return Ok(());
        }
        print_header(&provider.summarize(&raw)?);
        for stub in stubs {
            println!(
                "{} {} {}",
                format::pad(&stub.index.to_string(), 4),
                format::role_icon(stub.role),
                format::truncate(&stub.preview, 110)
            );
        }
    }
    Ok(())
}

fn print_header(summary: &SessionSummary) {
    println!("{}", summary.title);
    let mut meta = format!(
        "{} · {} · {} messages · {}",
        summary.id,
        format::abbreviate_path(&summary.project_path),
        summary.message_count,
        format::relative_time(&summary.mtime),
    );
    if let Some(branch) = &summary.git_branch {
        meta.push_str(&format!(" · {branch}"));
    }
    println!("{meta}");
    println!("{}", "─".repeat(80));
}

/// P1 extras carried in provider extensions: token totals, subagents,
/// artifacts.
fn print_extensions(session: &Session) {
    if let Some(usage) = session.extensions.get("usage") {
        let get = |key: &str| usage.get(key).and_then(|v| v.as_u64()).unwrap_or(0);
        println!(
            "tokens: in {} · out {} · cache-read {} · cache-write {}",
            get("inputTokens"),
            get("outputTokens"),
            get("cacheReadTokens"),
            get("cacheCreationTokens"),
        );
    }
    if let Some(subagents) = session
        .extensions
        .get("subagents")
        .and_then(|v| v.as_array())
    {
        for agent in subagents {
            let label = agent
                .get("description")
                .or(agent.get("agentId"))
                .and_then(|v| v.as_str())
                .unwrap_or("?");
            let count = agent
                .get("messageCount")
                .and_then(|v| v.as_u64())
                .unwrap_or(0);
            println!("subagent: {label} ({count} messages)");
        }
    }
    if let Some(artifacts) = session
        .extensions
        .get("artifacts")
        .and_then(|v| v.as_array())
    {
        for artifact in artifacts {
            let name = artifact.get("name").and_then(|v| v.as_str()).unwrap_or("?");
            println!("artifact: {name}");
        }
    }
}

fn print_full(message: &Message) {
    println!(
        "{} [{}]",
        format::role_icon(message.role),
        serde_json::to_value(message.role)
            .unwrap()
            .as_str()
            .unwrap_or("?")
    );
    for block in &message.blocks {
        match block {
            ContentBlock::Text { text } => println!("{text}"),
            ContentBlock::ToolUse { name, .. } => println!("  ⚙ tool_use: {name}"),
            ContentBlock::ToolResult { summary, truncated } => {
                let first_line = summary.lines().next().unwrap_or("");
                let ellipsis = if *truncated { " …" } else { "" };
                println!(
                    "  → tool_result: {}{ellipsis}",
                    format::truncate(first_line, 100)
                );
            }
            ContentBlock::Thinking { .. } => println!("  ✳ thinking (collapsed)"),
        }
    }
    println!();
}

/// ADR-4 CLI resume: chdir into the project, then exec the provider's resume
/// command, replacing this process. Never returns on success — except for
/// GUI-app sessions (Cursor), whose launcher returns immediately.
pub fn resume(store: &SessionStore, session_id: &str) -> Result<()> {
    let spec = store.resume_spec_for(session_id)?;
    let (program, args) = spec
        .command
        .split_first()
        .ok_or_else(|| Error::Data("Empty resume command".to_string()))?;
    if spec.launch == vibetrail_core::LaunchMode::GuiApp {
        let status = std::process::Command::new(program)
            .args(args)
            .current_dir(&spec.project_path)
            .status()
            .map_err(|e| Error::Data(format!("Failed to run {}: {e}", spec.command.join(" "))))?;
        if !status.success() {
            return Err(Error::Data(format!(
                "Launcher failed: {}",
                spec.command.join(" ")
            )));
        }
        println!("Opened the app at the project — pick this session up from its chat history.");
        return Ok(());
    }
    let error = std::process::Command::new(program)
        .args(args)
        .current_dir(&spec.project_path)
        .exec();
    // Only reached when exec failed.
    Err(Error::Data(format!(
        "Failed to exec {}: {error}",
        spec.command.join(" ")
    )))
}

/// `vibetrail config`: the effective settings — config file location, which
/// providers are enabled and which store roots are in effect.
pub fn config_report(json: bool) -> Result<()> {
    let report = vibetrail_core::config::report(&vibetrail_core::config::load_discovery());
    if json {
        println!("{}", to_json(&report)?);
        return Ok(());
    }
    println!("config file: {}", format::abbreviate_path(&report.path));
    for provider in &report.providers {
        let state = if provider.enabled {
            "enabled"
        } else {
            "disabled"
        };
        let mut notes = Vec::new();
        if provider.root_is_custom {
            notes.push("custom");
        }
        if !provider.root_exists {
            notes.push("missing");
        }
        let suffix = if notes.is_empty() {
            String::new()
        } else {
            format!("  ({})", notes.join(", "))
        };
        println!(
            "  {} {} {}{suffix}",
            format::pad(&provider.id, 13),
            format::pad(state, 9),
            format::abbreviate_path(&provider.root),
        );
    }
    Ok(())
}

pub fn open_gui(project: Option<&str>) -> Result<()> {
    let mut command = std::process::Command::new("/usr/bin/open");
    command.args(["-a", "VibeTrail"]);
    if let Some(project) = project {
        command.arg(project);
    }
    let status = command
        .status()
        .map_err(|e| Error::Data(format!("Failed to run open: {e}")))?;
    if !status.success() {
        return Err(Error::Data(
            "VibeTrail.app not found; build it with `cargo tauri build`".to_string(),
        ));
    }
    Ok(())
}
